core/src/main/java/tech/tablesaw/io/Source.java

```diff
@@ -1,121 +1,143 @@
 package tech.tablesaw.io;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.io.ByteStreams;
 import com.ibm.icu.text.CharsetDetector;
 import com.ibm.icu.text.CharsetMatch;
-import java.io.ByteArrayInputStream;
 import java.io.File;
-import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.io.RandomAccessFile;
 import java.io.Reader;
-import java.io.StringReader;
 import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel.MapMode;
 import java.nio.charset.Charset;
-import java.util.Scanner;
 
 public class Source {
 
-  // we always have one of these (file, reader, or inputStream)
-  protected final File file;
-  protected final Reader reader;
-  protected final InputStream inputStream;
-  protected final Charset charset;
+  protected final ByteBuffer byteBuffer;
+  protected Charset charset;
 
-  public Source(File file) {
-    this(file, getCharSet(file));
+  public Source(File file, Charset charset) throws IOException {
+    this(getMappedByteBuffer(file), charset);
   }
 
-  public Source(File file, Charset charset) {
-    this.file = file;
-    this.reader = null;
-    this.inputStream = null;
-    this.charset = charset;
+  public Source(File file) throws IOException {
+    this(getMappedByteBuffer(file));
+  }
+
+  private static MappedByteBuffer getMappedByteBuffer(File file) throws IOException {
+    try (RandomAccessFile raf = new RandomAccessFile(file, "r")) {
+      return raf.getChannel().map(MapMode.READ_ONLY, 0, raf.length());
+    }
+  }
+
+  public Source(InputStream inputStream, Charset charset) throws IOException {
+    this(ByteStreams.toByteArray(inputStream), charset);
```

Comment:
This will read all bytes from the stream into an array, which seems less efficient than streaming processing

Reply:
Do you mean memory or time efficiency? It would be good to measure the impact of this change. Is there a benchmark available for tablesaw?

But note that this is not different from the previous implementation. If you look at the previous code from `getReaderAndColumnTypes()`, which was duplicated in both `CsvReader` and `FixedWidthReader`, you'll see that it caches the whole stream into a byte array of any input source different from a file:
```
bytesCache = CharStreams.toString(reader).getBytes();
```
Indeed, this line crashes the JVM for files larger than ~2GB because it tries to allocate an array with an index larger than an int can hold. This happened to me while reading a file via a `FileInputStream`.

Now, InputStreams are read once to a byte buffer and reused whenever necessary later (this still has the previous problem). Files are read directly from a MappedByteBuffer, so they are directly read from the OS cache when necessary. In my opinion, this simplifies the logic from `Source` class and their consumers which do not need to keep track of the type of source and close the underlying resources when necessary.

Reply:
I believe that before it would only cache for those two file types if the user did not supply the column types and it looks like it will now happen for all types under all circumstances. Since you said the read ahead may crash if the input is larger than 2GB I think I'd rather avoid it happening in a wider variety of circumstances.

Reply:
> But note that this is not different from the previous implementation. If you look at the previous code from getReaderAndColumnTypes(), which was duplicated in both CsvReader and FixedWidthReader, you'll see that it caches the whole stream into a byte array of any input source different from a file:

hmm. I think @benmccann is correct. My understanding is that the reader _sometimes_ reads the whole file into memory, but that that isn't necessarily the case. Would this change mean always reading the file into memory?

> Indeed, this line causes the JVM to throw OOM for files larger than ~2GB because it tries to allocate an array with an index larger than an int can hold. 

I'm not sure I'm getting this. Where is there an index that's based on the file size? I know we would throw an error for more than Integer.max _rows_, but wasn't aware of a file size issue.

Reply:
> hmm. I think @benmccann is correct. My understanding is that the reader _sometimes_ reads the whole file into memory, but that that isn't necessarily the case.

Yeah, my understanding is also that the reader only caches the whole file sometimes. More specifically, it caches when an `InputStream` or a `Source`  is provided as input instead of a `File`:
```
      if (options.source().file() == null) {	
        bytesCache = CharStreams.toString(reader).getBytes();	
        // create a new reader since we just exhausted the existing one	
        reader = options.source().createReader(bytesCache);	
      }
```

> Would this change mean always reading the file into memory?

No. This only reads the file into memory in the same situations as before. When an `InputStream` or a `Source` is provided, the data is cached in a byte array wrapped in a `ByteBuffer`. When a `File` is provided, the file is read directly using memory mapping via a `MappedByteBuffer`. This allows the rest of the code to always deal with the same interface.



> I'm not sure I'm getting this. Where is there an index that's based on the file size? I know we would throw an error for more than Integer.max _rows_, but wasn't aware of a file size issue.

In the cases where caching happens, the method `CharStreams.toString(reader)` reads the whole file into a single `String`, and strings have a maximum size due to array indexing issues mentioned before.

If I recall correctly, I got an OOM error when reading a large file using a `FileInputStream`, which forces `tablesaw` to read and cache the entire file. Debugging this issue motivated me to create this pull request. Only later I realized that providing a `File` instance was enough to solve my problem.

Reply:
> it caches when an InputStream or a Source is provided as input instead of a File

I don't think that's quite right. It will cache it when an `InputStream` is provided, but I don't think it necessarily will when a `Source` is provided. A `Source` is just a container for some type of input. If the underlying input for the `Source` is a `File` then it won't read it into memory and cache it

Reply:
You're right, what I meant was a `Reader`, not a `Source`. I updated the comment to reflect that.

Reply:
Additionally, the caching happens immediately when the `Source` container is created, which is why the constructors had to change to throw an `IOException`.